From 5bc41e6b2bf2510d89c373cd8d54e3ecf56c090f Mon Sep 17 00:00:00 2001
From: Guillem Jover <guillem@hadrons.org>
Date: Mon, 25 Jan 2010 17:51:17 +0100
Subject: [PATCH 02/02] Fix segfaults due to mmap failure

If mmap fails the returned address is going to be invalid which is
going to make subsequent usege of the instance use bogus values causing
segfaults, handle this case as a normal open error.
---
 libfstt/rafile.cc |    5 +++++
 src/xfstt.cc      |   15 +++++++++++++++
 2 files changed, 20 insertions(+), 0 deletions(-)

diff --git a/libfstt/rafile.cc b/libfstt/rafile.cc
index 7155fcb..476e228 100644
--- a/libfstt/rafile.cc
+++ b/libfstt/rafile.cc
@@ -184,6 +184,11 @@ RandomAccessFile::RandomAccessFile(char
 	length = st.st_size;
 	base = (u8_t *)mmap(0L, length, PROT_READ, MAP_SHARED, fd, 0L);
 	close(fd);
+	if (base == MAP_FAILED) {
+		debug("MMap failed '%s'\n", strerror(errno));
+		ptr = absbase = base = 0;
+		return;
+	}
 	ptr = absbase = base;
 }
 
diff --git a/src/xfstt.cc b/src/xfstt.cc
index df394b7..d02f2d9 100644
--- a/src/xfstt.cc
+++ b/src/xfstt.cc
@@ -779,6 +779,11 @@ openTTFdb()
 	close(fd);
 	delete ttinfofilename;
 
+	if (infoBase == MAP_FAILED) {
+		error(_("cannot mmap font database!\n"));
+		return 0;
+	}
+
 	if (infoSize <= sizeof(TTFNheader)
 	    || strncmp(infoBase, "TTFNINFO", 8)) {
 		error(_("corrupt font database!\n"));
@@ -807,6 +812,11 @@ openTTFdb()
 	close(fd);
 	delete ttnamefilename;
 
+	if (nameBase == MAP_FAILED) {
+		error(_("cannot mmap font database!\n"));
+		return 0;
+	}
+
 	if (nameSize <= sizeof(TTFNheader)
 	    || strncmp(nameBase, "TTFNNAME", 8)) {
 		error(_("corrupt font database!\n"));
@@ -826,6 +836,11 @@ openTTFdb()
 		aliasBase = (char *)mmap(0L, aliasSize, PROT_READ, MAP_SHARED,
 					 fd, 0L);
 		close(fd);
+
+		if (aliasBase == MAP_FAILED) {
+			error(_("cannot mmap alias database!\n"));
+			return 0;
+		}
 	}
 
 	return 1;
-- 
1.7.1

